### PR TITLE
*: hook up new persist-txn tables implementation behind a flag

### PR DESCRIPTION
**Pull request by danhhz**
This is primarily in service of Pv2's use-case isolation. Switching our tables
to persist-txn is necessary, but not sufficient work to tease apart table txns
into isolated units.

The persist-txn library presents an abstraction on top of persist shards,
allowing atomic multi-shard writes. This is accomplished through an additional
_txns_ shard that coordinates writes to a (potentially large) number of _data_
shards. Data shards may be added to and removed from the set at any time.

From Pv2's perspective the motivating benefit is that these multi-shard txns are
_atomic_. This will allow us to build an Optimistic Concurrency Control protocol
for our SQL transactions, removing the single-process global write lock.

Eventual secondary benefits of persist-txn tables will be:
- Greatly decreased Consensus/CRDB write traffic. Currently, tables require
  `O(num_tables * passage_of_time)` writes. With persist_txn, that will be
  drastically reduced to `O(num_tables_in_txn) + O(passage_of_time)`. That is,
  before we had to write to each table once per second, after we will write one
  write per second plus one for each table involved in a txn. This will
  hopefully allow us to essentially remove the system limit on the number of
  tables.
- Bounded memory table writes. Technically, this is possible today, but the
  division of responsibilities between adaptor and storage makes it tricky. The
  persist-txn API will make it much easier to confidently surface lower-level
  table details up to adaptor.

This PR is intended to have no behavior difference when the flag is off.

Touches #22173

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

I was really hoping to get this branch to a point where I could merge it with
the new behavior on-by-default in CI, but a number of work items have turned out
be better handled as post-integration followups. These are tracked in #22801.
Notably, many of these are performance related (crdb writes are down, but crdb and
s3 reads are up and a couple extra ops in the INSERT path) and we haven't finished
fleshing out the story for flipping the feature flag off. There are also a few
last CI issues I'm working through.

However, if we like, this should be enough to start writing tests for flipping
the feature on and off (restarting every process for each transition). This
likely has some correctness issues at the moment, which might make it a good
sanity check that new tests are tickling interesting edge cases.

As of this PR, the flag should enable us to have far more tables than previously
without knocking over crdb, but the above outstanding performance issues mean
that overall mz performance is down until more of the followups are done.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/persist-txn/src/operator.rs
Comment (philip-stoev) [inline src/persist-txn/src/operator.rs:142]: This triggered in the Nightly Canary CI step and ended up in Sentry:

https://materializeinc.sentry.io/issues/4386718273/?project=6780145

**Comment by philip-stoev** [comment]
A second panic happened on Sentry with this commit:

https://materializeinc.sentry.io/issues/4386716466/

`TODO` panic in `mz_persist_types::codec_impls::TodoSchema<T>::columns (codec_impls.rs:776)`

**Comment by danhhz** [comment]
ah, sorry! this was very experimental and I just wanted to get a sense for what sorts of failures that I had left to work through. i'll comment out the staging canary in this branch

File: src/environmentd/src/bin/environmentd/main.rs
Comment (danhhz) [inline src/environmentd/src/bin/environmentd/main.rs:427]: note to self: we should plumb this to clusterd processes so that everything restarts when the flag is flipped on envd (as is necessary)

File: src/environmentd/src/bin/environmentd/main.rs
Comment (danhhz) [inline src/environmentd/src/bin/environmentd/main.rs:904]: note to self: hook to up to the flag before merge (once done iterating on CI with the flag on)

**Comment by danhhz** [comment]
@jkosh44 This should be ready for a look! As I mention in the PR description, I'm sadly going to have to leave some things for followup work, including flipping CI to have the flag enabled. We'll also certainly want to get someone from storage to look at it, I might go poke around in the storage hangout tomorrow to see if I can find someone.

File: src/environmentd/src/bin/environmentd/main.rs
Comment (jkosh44) [reply src/environmentd/src/bin/environmentd/main.rs:904]: Just to confirm, the plan is to update this to `enable_persist_txn_tables: args.enable_persist_txn_tables` before merging? I'm going to hold off on an explicit approval until it's changed just to be extra sure we don't accidentally merge as is.

**Review by jkosh44 — COMMENTED** [review]
LGTM, holding off on explicit approval until the flag is updated.

File: src/environmentd/src/bin/environmentd/main.rs
Comment (danhhz) [reply src/environmentd/src/bin/environmentd/main.rs:904]: Yup, that's the plan! I've started working on getting CI to fully pass in some followups, so I'll go ahead and push this change now

**Review by jkosh44 — APPROVED** [review]

**Review by petrosagg — APPROVED** [review]
I focused my review primarily on making sure nothing breaks subtly with the feature flag is off and secondarily on going through the design doc and reviewing the transaction work.

When the flag is off everything looks sensible and I think is safe to merge and deploy to production. When the flag is on I followed most of the changes and had a couple of questions, mostly for my education, but it looks like more iterations are planned on this code. Let me know if this is not the case and the flag is intended to be turned on in any prod-like environment with the code in the current state.

Beyond this PR, my main question and concern about this transaction implementation is that there is no synchronization between registering a shard to the transaction shard and stopping all direct writes to it. I think the registration of a shard to a transaction shard needs to be primarily an operation on the shard itself, so that any future direct writers can know that this shard is no longer available for direct writes, and then lazily registered to the transaction shard (so the WAL direction is inverted in this case). This is a bit tricky when both the data and tx shards share a timestamp type but pretty easy when they're not.

File: src/storage-controller/src/lib.rs
Comment (danhhz) [reply src/storage-controller/src/lib.rs:1207]: Added a comment, does this help clarify?

Consider the following scenario:
- Table A is written to via txns at time 5
- Tables other than A are written to via txns consuming timestamps up to 10
- We'd like to read A at 7
- The application process of A's txn has advanced the upper to 5+1, but we need it to be past 7, but the txns shard knows that (5,10) is empty of writes to A
- This branch allows it to handle that advancing the physical upper of Table A to 10 (NB but only once we see it get past the write at 5!)
- Then we can read it normally.

File: src/storage-controller/src/lib.rs
Comment (danhhz) [reply src/storage-controller/src/lib.rs:2002]: Yeah, this is an awkward artifact that's one of my followups (it's a `TODO(txn)` somewhere).

To write a batch in persist, you need a real schema, but to `compare_and_append_batch` with something already written out to s3, you don't. There's some internal plumbing bits where we don't have a real WriteHandle but need to this `compare_and_append_batch` of an already written batch (i.e. application of an already committed txn).

The fix will be to split out a SchemalessWriteHandle (name TBD) that can do things like `compare_and_append_batch` but nothing that requires a schema.

**Review by danhhz — COMMENTED** [review]
> I focused my review primarily on making sure nothing breaks subtly with the feature flag is off and secondarily on going through the design doc and reviewing the transaction work.

Perfect, I was a little stressed about this. Very happy to have extra eyes on it!

> When the flag is off everything looks sensible and I think is safe to merge and deploy to production. When the flag is on I followed most of the changes and had a couple of questions, mostly for my education, but it looks like more iterations are planned on this code. Let me know if this is not the case and the flag is intended to be turned on in any prod-like environment with the code in the current state.

That is entirely correct. Before turning this on in prod, I'm going to address everything in #22801 (most of that is the `TODO(txn)` catchall) and QA is going to get to kick the tires.

> Beyond this PR, my main question and concern about this transaction implementation is that there is no synchronization between registering a shard to the transaction shard and stopping all direct writes to it. I think the registration of a shard to a transaction shard needs to be primarily an operation on the shard itself, so that any future direct writers can know that this shard is no longer available for direct writes, and then lazily registered to the transaction shard (so the WAL direction is inverted in this case). This is a bit tricky when both the data and tx shards share a timestamp type but pretty easy when they're not.

Started discussing this offline and discovered that Petros had a mental model of us pulling data shards in and out of the txns shard for things like fast uncontended writes when there is no need for transactions (in which case, yeah the above is super scary). I'm hoping we don't need to do something like that (in which case much less scary) and am talking him through my logic. Discussions ongoing

**Comment by danhhz** [comment]
Okay, I've ended up in rebase hell again, so gonna go ahead and get this merged. Petros and I have a followup scheduled for tomorrow about his outstanding questions

TFTRs!!

**Pull request merged**
